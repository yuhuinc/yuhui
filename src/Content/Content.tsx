import React, { createContext, useContext, useState } from "react";
import styled, { css } from "styled-components";

export const ContentContext = createContext<ContentContext>({
  lang: "",
  setLang: () => {},
  theme: {},
  contentNodes: {},
  addContentNodes: () => {}
});

interface Copy {
  en: string;
  fr: string;
  [key: string]: string;
}

interface ContentNode {
  copy: Copy;
  style?: any;
  render?: any;
  [key: string]: any;
}

interface ContentNodes {
  [key: string]: ContentNode;
}

interface Content {
  contentNodes: ContentNodes;
  theme: any;
}

interface ContentContext {
  lang: string;
  setLang: (lang: string) => void;
  contentNodes: ContentNodes;
  addContentNodes: (nodes: ContentNodes) => void;
  theme: any;
}

interface ContentProviderProps {
  children?: React.ReactNode;
  lang?: string;
  content: Content;
  [key: string]: any;
}

export const ContentProvider = ({
  children,
  lang: defaultLang,
  content
}: ContentProviderProps) => {
  const { contentNodes: defaultContentNodes, theme } = content;
  const [lang, setLang] = useState(defaultLang || "en");
  const [contentNodes, setContentNodes] = useState(
    defaultContentNodes || ({} as ContentNodes)
  );

  const addContentNodes = (nodes: ContentNodes) => {
    setContentNodes({ ...contentNodes, ...nodes });
  };

  return (
    <ContentContext.Provider
      value={{
        lang,
        setLang,
        contentNodes,
        addContentNodes,
        theme
      }}
    >
      {children}
    </ContentContext.Provider>
  );
};

export function interpolate(string, params) {
  return string.split(/\${|}/gm).reduce((prev, curr, i) => {
    if (i % 2 === 0) {
      return prev + curr;
    } else {
      return prev + params[curr];
    }
  }, "");
}

export interface WithContentProps {
  contentKey: string;
  copyParams?: any;
  style?: any;
  render?: any;
  [key: string]: any;
}

export const withContent = (WrappedComponent: React.ComponentType<any>) => ({
  style,
  contentKey,
  copyParams = {},
  render,
  children,
  ...rest
}: WithContentProps) => {
  const { contentNodes, theme } = useContext(ContentContext);
  const copy = useCopy(contentKey, copyParams);
  return (
    <WrappedComponent
      style={{ ...contentNodes?.[contentKey]?.style, ...style }}
      theme={theme}
      {...rest}
    >
      <>
        {render ? render(copy) : copy}
        {children}
      </>
    </WrappedComponent>
  );
};

export const useContent = (): ContentContext => {
  return useContext(ContentContext);
};

export const useCopy = (contentKey: string, copyParams?) => {
  const { lang, contentNodes } = useContent();
  const copy = contentNodes?.[contentKey]?.copy?.[lang];
  if (!copy) {
    return `MISSING translations for: ${contentKey}`;
  } else if (copyParams) {
    return interpolate(copy, copyParams);
  } else {
    return copy;
  }
};

const baseTextStyles = css`
  ${props =>
    props.centered &&
    `
    text-align: center;
  `};
`;
export const H1 = withContent(styled.h1`
  ${baseTextStyles};
`);
export const H2 = withContent(styled.h2`
  ${baseTextStyles};
`);
export const P = withContent(styled.p`
  ${baseTextStyles};
`);
export const Span = withContent(styled.span`
  ${baseTextStyles};
`);
