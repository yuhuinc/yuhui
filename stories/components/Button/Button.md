# Buttons

Accessible buttons.

**Available components**:

- [ButtonPrimary](#buttonprimary)

**How it is built**

- Uses `reakit/Button` ([link](https://reakit.io/docs/button/)) under the hood.
- Wrapped in `withContent`.

## ButtonPrimary

### Example

```javascript
<ButtonPrimary contentKey="demo.Button" onClick={() => setLang("en")} />
```
