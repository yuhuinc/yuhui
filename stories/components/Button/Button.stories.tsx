import React from "react";
import PropTypes from "prop-types";

import { ContentProvider, useContent, ButtonPrimary } from "../../../src";

export default {
  title: "Components/Button"
};

const contentNodes = {
  "demo.Button": {
    copy: {
      en: "Switch to English",
      fr: "Passer à l'anglais"
    }
  },
  "demo.ButtonWithRender": {
    copy: {
      en: "Switch to French",
      fr: "Passer au français"
    }
  }
};

const theme = {
  colors: {
    primary: {
      light: "#d0dbfd",
      dark: "#230cc2"
    }
  }
};

const content = {
  contentNodes,
  theme
};

/**
 * Use `Badge` to highlight key info with a predefined status.
 */
export const PrimaryButton = () => {
  return (
    <ContentProvider lang="fr" content={content}>
      <Buttons />
    </ContentProvider>
  );
};
PrimaryButton.propTypes = {
  test: PropTypes.string
};

const Buttons = () => {
  const { lang, setLang } = useContent();

  return (
    <div>
      <p>Language code: {lang}</p>
      <ButtonPrimary contentKey="demo.Button" onClick={() => setLang("en")} />
      <ButtonPrimary
        onClick={() => setLang("fr")}
        contentKey="demo.ButtonWithRender"
        render={({ content, lang }) => (
          <span>
            <b>{content.copy[lang]}</b> with render()
          </span>
        )}
      />
    </div>
  );
};
