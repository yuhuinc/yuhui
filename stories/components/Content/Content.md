# Content

Functions exported from `Content.tsx` is in large part what makes this component library what it is. Many of the components in this library is wrapped in `useContent()` which gets its content (copy and styles) from `<ContentProvider />`

**Available components**:

- [ContentProvider](#contentprovider)
- [withContent](#withcontent)
- [Basic text components](#basic-text-components)

**How it is built**

- No use of external component libraries.
- Uses React's `context` under the hood.

## ContentProvider

### Props

| name     | type      | default | description                    |
| -------- | --------- | ------- | ------------------------------ |
| lang     | string    | "en"    | Language code                  |
| content  | Content   |         | **Required.** `Content` object |
| children | ReactNode |         |                                |

### Types

```javascript
interface Content {
  contentNodes: ContentNodes;
  theme: any;
}

interface ContentNodes {
  [key: string]: ContentNode;
}

interface ContentNode {
  copy: Copy;
  style?: any;
  render?: any;
  [key: string]: any;
}
```

### Example

```javascript
const content = {
  contentNodes: {
      {
  "demo.Button": {
    copy: {
      en: "Switch to English",
      fr: "Passer Ã  l'anglais"
    }
  }
};
  },
  theme: {
    colors: {
      primary: {
        light: "#d0dbfd",
        dark: "#230cc2"
      }
    }
  }
};

...

<ContentProvider lang="en" content={content}>
    ...
</ContentProvider>
```

## withContent

A higher-order component which adds the following props.

### Props

| name       | type            | default | description                                                                  |
| ---------- | --------------- | ------- | ---------------------------------------------------------------------------- |
| contentKey | string          |         | **Required.** Name of the corresponding key in `contentNodes` from `Content` |
| style      | Style           |         | Style object. Overwrites styles for the root element.                        |
| render     | () => ReactNode |         | A render prop that exposes `content` and `lang`.                             |
| children   | ReactNode       |         | An escape hatch for rendering content that is not part of `Content`.         |

### Examples

```javascript
const H1 = withContent(({children, ...rest}) => <h1 {...rest}>{children}</h1>);


// Below will render copy from content.contentNodes['demo.Button']
<H1 contentKey="demo.Button" />

// You can have more control with the render prop
<H1
    render={({content, lang}) => `${content[lang]} !!`}
/>
```

## Basic Text Components

Basic text components are all wrapped with `withContent`

- `H1`
- `H2`
- `P`
- `Span`
