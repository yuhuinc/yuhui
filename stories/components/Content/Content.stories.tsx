import React from "react";

import markdown from "./Content.md";
import { ContentProvider, H1 } from "../../../src";

export default {
  title: "Components/Content"
};

const contentNodes = {
  "demo.H1": {
    copy: {
      en: 'View docs by clicking "Notes"',
      fr: 'View docs by clicking "Notes"'
    }
  }
};

const theme = {
  colors: {
    primary: {
      light: "#d0dbfd",
      dark: "#230cc2"
    }
  }
};

const content = {
  contentNodes,
  theme
};

export const Content = () => {
  return (
    <ContentProvider lang="fr" content={content}>
      <H1 contentKey="demo.H1" />
    </ContentProvider>
  );
};
Content.story = {
  parameters: {
    notes: { markdown }
  }
};
