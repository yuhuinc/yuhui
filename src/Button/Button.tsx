import React from "react";
import { Button as RKButton } from "reakit/Button";

import { withContent, WithContentProps } from "../Content/Content";

export interface ButtonProps extends WithContentProps {
  disabled?: boolean;
  [key: string]: any;
}

export type Button = React.ComponentType<ButtonProps>;

export const Button: Button = withContent(
  ({ children, ...rest }) => {
    return <RKButton {...rest}>{children}</RKButton>;
  }
);
