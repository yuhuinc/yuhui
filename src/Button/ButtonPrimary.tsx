import React from "react";
import styled from "styled-components";

import { colors } from "../shared/constants";
import { withContent } from "../Content/Content";
import { ButtonProps } from "./Button";

const StyledButton = styled.button`
  color: ${props =>
    props.disabled
      ? colors.BATTLESHIP_GREY
      : props.theme?.colors?.primary?.dark || colors.PURE_BLUE};
  background-color: ${props =>
    props.disabled
      ? colors.DISABLED_GRAY
      : props.theme?.colors?.primary?.light || colors.BLUEY_GREY};
  line-height: 36px;
  height: 36px;
  font-size: 14px;
  font-weight: bold;
  border: none;
  border-radius: 28px !important;
  padding: 1px 24px 0;
  text-align: center;
  cursor: ${props => (props.disabled ? "not-allowed" : "pointer")};
`;

export const ButtonPrimary = withContent(
  ({ children, theme, ...rest }: ButtonProps) => {
    return (
      <StyledButton {...rest} theme={theme}>
        {children}
      </StyledButton>
    );
  }
);
