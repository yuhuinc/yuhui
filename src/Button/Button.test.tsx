import React from "react";
import faker from "faker";
import { render, fireEvent } from "@testing-library/react";
import { Button } from "./Button";
import { ContentProvider } from "../Content/Content";

const TestButton = ({
  contentNodes = {},
  theme = {},
  lang = "en",
  contentKey,
  ...rest
}) => {
  const defaultContentNodes = {};

  const defaultTheme = {
    colors: {
      primary: {
        light: "#d0dbfd",
        dark: "#230cc2"
      }
    }
  };

  return (
    <ContentProvider
      content={{
        theme: { ...defaultTheme, ...theme },
        contentNodes: { ...defaultContentNodes, ...contentNodes }
      }}
      lang={lang}
    >
      <Button contentKey={contentKey} {...rest} />
    </ContentProvider>
  );
};

describe("content", () => {
  test("renders English copy", () => {
    const enCopy = "snapshot";
    const contentNodes = {
      button: {
        copy: {
          en: enCopy,
          fr: faker.lorem.word()
        }
      }
    };
    const { container } = render(
      <TestButton lang="en" contentNodes={contentNodes} contentKey="button" />
    );

    expect(container).toMatchSnapshot();
  });

  test("renders French copy", () => {
    const frCopy = "snapshot";
    const contentNodes = {
      button: {
        copy: {
          en: faker.lorem.word(),
          fr: frCopy
        }
      }
    };
    const { container } = render(
      <TestButton lang="fr" contentNodes={contentNodes} contentKey="button" />
    );

    expect(container).toMatchSnapshot();
  });
});

describe("events", () => {
  test("calls onClick callback when click event is fired", async () => {
    const handleClick = jest.fn(() => {});
    const { getByTestId } = render(
      <TestButton
        data-testid="button"
        onClick={handleClick}
        contentKey="hello"
      />
    );

    fireEvent.click(getByTestId("button"));

    expect(handleClick.mock.calls.length).toBe(1);
  });
});
