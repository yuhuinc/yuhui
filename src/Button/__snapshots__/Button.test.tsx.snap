// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content renders English copy 1`] = `
<div>
  <button
    theme="[object Object]"
    type="button"
  >
    snapshot
  </button>
</div>
`;

exports[`content renders French copy 1`] = `
<div>
  <button
    theme="[object Object]"
    type="button"
  >
    snapshot
  </button>
</div>
`;
